File: mobile-rn/src/screens/ChatScreen.js
import { useEffect, useLayoutEffect, useRef, useState } from "react";
import { FlatList, KeyboardAvoidingView, Platform, Pressable, StyleSheet, Text, View } from "react-native";
import { colors } from "../theme";
import { Field } from "../components/ui";
import { api } from "../api";
import { useAuth } from "../context/AuthContext";
import { getBuyerId } from "../context/AuthContext";

function timeAgo(ts) {
  const s = Math.floor((Date.now() - ts) / 1000);
  if (s < 60) return "à l'instant";
  const m = Math.floor(s / 60);
  if (m < 60) return `il y a ${m} min`;
  const h = Math.floor(m / 60);
  if (h < 24) return `il y a ${h} h`;
  return `il y a ${Math.floor(h / 24)} j`;
}

export default function ChatScreen({ route, navigation }) {
  const { product } = route.params;
  const { auth } = useAuth();
  const [messages, setMessages] = useState([]);
  const [input, setInput] = useState("");
  const [mySenderId, setMySenderId] = useState(auth?.phone || null);
  const listRef = useRef(null);

  useLayoutEffect(() => {
    navigation.setOptions({ title: product.name });
  }, [navigation, product.name]);

  useEffect(() => {
    let mounted = true;
    if (!auth) getBuyerId().then((id) => mounted && setMySenderId(id));
    return () => {
      mounted = false;
    };
  }, [auth]);

  useEffect(() => {
    let cancelled = false;
    async function poll() {
      try {
        const data = await api.chat(product.id);
        if (!cancelled) setMessages(data.messages);
      } catch {
        /* silent — polling, no need to surface transient errors */
      }
    }
    poll();
    const timer = setInterval(poll, 2500);
    return () => {
      cancelled = true;
      clearInterval(timer);
    };
  }, [product.id]);

  const isSeller = mySenderId === product.sellerPhone;

  async function handleSend() {
    const text = input.trim();
    if (!text || !mySenderId) return;
    setInput("");
    try {
      await api.sendChat(product.id, mySenderId, text);
      const data = await api.chat(product.id);
      setMessages(data.messages);
    } catch {
      /* if it fails, the message just won't appear — user can retry */
    }
  }

  return (
    <KeyboardAvoidingView style={styles.flex} behavior={Platform.OS === "ios" ? "padding" : undefined} keyboardVerticalOffset={90}>
      <Text style={styles.vendorLabel}>{isSeller ? "Vous discutez en tant que vendeur" : product.vendorId}</Text>
      <FlatList
        ref={listRef}
        data={messages}
        keyExtractor={(m) => m.id}
        contentContainerStyle={styles.list}
        onContentSizeChange={() => listRef.current?.scrollToEnd({ animated: true })}
        renderItem={({ item }) => {
          const mine = item.senderId === mySenderId;
          return (
            <View style={[styles.msgRow, mine ? styles.msgRowMine : styles.msgRowTheirs]}>
              <View style={[styles.bubble, mine ? styles.bubbleMine : styles.bubbleTheirs]}>
                <Text style={mine ? styles.bubbleTextMine : styles.bubbleText}>{item.text}</Text>
              </View>
              <Text style={styles.meta}>
                {mine ? "Vous" : item.role === "seller" ? "Vendeur" : item.senderId} · {timeAgo(item.createdAt)}
              </Text>
            </View>
          );
        }}
      />
      <View style={styles.inputRow}>
        <Field style={styles.input} placeholder="Écrire un message…" value={input} onChangeText={setInput} onSubmitEditing={handleSend} />
        <Pressable style={styles.sendBtn} onPress={handleSend}>
          <Text style={styles.sendBtnText}>➤</Text>
        </Pressable>
      </View>
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
  flex: { flex: 1, backgroundColor: colors.ink },
  vendorLabel: { color: colors.teal, fontSize: 11, paddingHorizontal: 16, paddingTop: 10 },
  list: { padding: 16, gap: 12 },
  msgRow: { maxWidth: "80%" },
  msgRowMine: { alignSelf: "flex-end", alignItems: "flex-end" },
  msgRowTheirs: { alignSelf: "flex-start", alignItems: "flex-start" },
  bubble: { borderRadius: 14, paddingHorizontal: 13, paddingVertical: 9 },
  bubbleMine: { backgroundColor: colors.amber, borderBottomRightRadius: 4 },
  bubbleTheirs: { backgroundColor: colors.panelRaised, borderWidth: 1, borderColor: colors.hairline, borderBottomLeftRadius: 4 },
  bubbleText: { color: colors.text, fontSize: 13 },
  bubbleTextMine: { color: colors.amberOn, fontSize: 13 },
  meta: { color: colors.textDim, fontSize: 9.5, marginTop: 4, paddingHorizontal: 4 },
  inputRow: { flexDirection: "row", gap: 8, padding: 12, borderTopWidth: 1, borderTopColor: colors.hairline, backgroundColor: colors.panelRaised },
  input: { flex: 1, borderRadius: 20 },
  sendBtn: { width: 40, height: 40, borderRadius: 20, backgroundColor: colors.amber, alignItems: "center", justifyContent: "center" },
  sendBtnText: { color: colors.amberOn, fontSize: 15 }
});

File: mobile-flutter/lib/screens/chat_screen.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../api.dart';
import '../auth.dart';
import '../theme.dart';
import '../time_ago.dart';
import '../widgets/app_field.dart';

class ChatScreen extends StatefulWidget {
  final Product product;

  const ChatScreen({super.key, required this.product});

  @override
  State<ChatScreen> createState() => _ChatScreenState();
}

class _ChatScreenState extends State<ChatScreen> {
  final _inputController = TextEditingController();
  final _scrollController = ScrollController();
  List<ChatMessage> _messages = [];
  String? _mySenderId;
  Timer? _timer;

  @override
  void initState() {
    super.initState();
    _mySenderId = context.read<AuthProvider>().auth?.phone;
    _resolveSenderId();
    _poll();
    _timer = Timer.periodic(const Duration(milliseconds: 2500), (_) => _poll());
  }

  Future<void> _resolveSenderId() async {
    if (_mySenderId == null) {
      final id = await getBuyerId();
      if (mounted) setState(() => _mySenderId = id);
    }
  }

  @override
  void dispose() {
    _timer?.cancel();
    _inputController.dispose();
    _scrollController.dispose();
    super.dispose();
  }

  Future<void> _poll() async {
    try {
      final thread = await Api.chat(widget.product.id);
      if (mounted) {
        setState(() => _messages = thread.messages);
        WidgetsBinding.instance.addPostFrameCallback((_) {
          if (_scrollController.hasClients) {
            _scrollController.animateTo(_scrollController.position.maxScrollExtent, duration: const Duration(milliseconds: 200), curve: Curves.easeOut);
          }
        });
      }
    } catch (_) {
      // polling — a transient failure just waits for the next tick
    }
  }

  Future<void> _send() async {
    final text = _inputController.text.trim();
    if (text.isEmpty || _mySenderId == null) return;
    _inputController.clear();
    try {
      await Api.sendChat(widget.product.id, _mySenderId!, text);
      await _poll();
    } catch (_) {
      // if it fails, the message just won't appear — user can retry
    }
  }

  @override
  Widget build(BuildContext context) {
    final isSeller = _mySenderId == widget.product.sellerPhone;

    return Scaffold(
      appBar: AppBar(title: Text(widget.product.name)),
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.fromLTRB(16, 10, 16, 0),
            child: Align(
              alignment: Alignment.centerLeft,
              child: Text(
                isSeller ? 'Vous discutez en tant que vendeur' : widget.product.vendorId,
                style: const TextStyle(color: AppColors.teal, fontSize: 11),
              ),
            ),
          ),
          Expanded(
            child: ListView.builder(
              controller: _scrollController,
              padding: const EdgeInsets.all(16),
              itemCount: _messages.length,
              itemBuilder: (context, index) {
                final m = _messages[index];
                final mine = m.senderId == _mySenderId;
                return Padding(
                  padding: const EdgeInsets.only(bottom: 12),
                  child: Column(
                    crossAxisAlignment: mine ? CrossAxisAlignment.end : CrossAxisAlignment.start,
                    children: [
                      Container(
                        constraints: BoxConstraints(maxWidth: MediaQuery.of(context).size.width * 0.8),
                        padding: const EdgeInsets.symmetric(horizontal: 13, vertical: 9),
                        decoration: BoxDecoration(
                          color: mine ? AppColors.amber : AppColors.panelRaised,
                          border: mine ? null : Border.all(color: AppColors.hairline),
                          borderRadius: BorderRadius.only(
                            topLeft: const Radius.circular(14),
                            topRight: const Radius.circular(14),
                            bottomLeft: Radius.circular(mine ? 14 : 4),
                            bottomRight: Radius.circular(mine ? 4 : 14),
                          ),
                        ),
                        child: Text(m.text, style: TextStyle(fontSize: 13, color: mine ? AppColors.amberOn : AppColors.text)),
                      ),
                      const SizedBox(height: 4),
                      Text(
                        '${mine ? 'Vous' : (m.role == 'seller' ? 'Vendeur' : m.senderId)} · ${timeAgo(m.createdAt)}',
                        style: const TextStyle(fontSize: 9.5, color: AppColors.textDim),
                      ),
                    ],
                  ),
                );
              },
            ),
          ),
          Container(
            padding: const EdgeInsets.all(12),
            decoration: const BoxDecoration(color: AppColors.panelRaised, border: Border(top: BorderSide(color: AppColors.hairline))),
            child: SafeArea(
              top: false,
              child: Row(
                children: [
                  Expanded(child: AppField(placeholder: 'Écrire un message…', controller: _inputController, onSubmitted: (_) => _send())),
                  const SizedBox(width: 8),
                  InkWell(
                    onTap: _send,
                    borderRadius: BorderRadius.circular(20),
                    child: Container(
                      width: 40,
                      height: 40,
                      alignment: Alignment.center,
                      decoration: const BoxDecoration(color: AppColors.amber, shape: BoxShape.circle),
                      child: const Text('➤', style: TextStyle(color: AppColors.amberOn, fontSize: 15)),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}
